addons/mail/models/mail_template.py

```diff
@@ -390,7 +390,11 @@ def generate_recipients(self, results, res_ids):
             if self._context.get('tpl_partners_only'):
                 mails = tools.email_split(values.pop('email_to', '')) + tools.email_split(values.pop('email_cc', ''))
                 for mail in mails:
-                    partner_id = self.env['res.partner'].find_or_create(mail)
+                    Partner = self.env['res.partner']
+                    if self.model and 'company_id' in self.env[self.model]._fields:
+                        company_id = self.env[self.model].browse(res_id).company_id.id
```

Comment:
I think that by browsing then reading in a loop you will have a SELECT by res_id, it would be better to do that outside of the loop.